### PR TITLE
Create automated app screenshots - Submission flow QR-Code/TAN (EXPOSUREAPP-2997)

### PR DESCRIPTION
**Pull request by jurajkusnier**
screenshots for submission flow

### Testing:

~~After targeting Android 11, MockK fails to run on API 11 when spy is used. For now please run the tests on pre-Android 11
mockk/mockk#519~~ (No longer necessary after target API downgrade)

1. install [fastlane](https://docs.fastlane.tools/getting-started/ios/setup/)
1. run `./gradlew clean assembleDebug assembleAndroidTest` for clean state
1. run `bundle exec fastlane screengrab` to execute tests

### CODE REVIEW

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionSymptomIntroFragmentTest.kt [hunk 1]
```diff
@@ -10,24 +11,54 @@ import androidx.test.ext.junit.runners.AndroidJUnit4
...
+    val systemUIDemoModeRule = SystemUIDemoModeRule()
+
+    private lateinit var viewModel: SubmissionSymptomIntroductionViewModel
 
     @Before
     fun setup() {
         MockKAnnotations.init(this, relaxed = true)
+        every { dispatcherProvider.Default } returns Dispatchers.Default
+        viewModel =
+            spyk(SubmissionSymptomIntroductionViewModel(dispatcherProvider, submissionRepository, autoSubmission))
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionSymptomIntroFragmentTest.kt:58]: you can use `TestDispatcherProvider`. no need to mock it

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultAvailableFragmentTest.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    val localeTestRule = LocaleTestRule()
+
+    @get:Rule
+    val systemUIDemoModeRule = SystemUIDemoModeRule()
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this, relaxed = true)
+
+        every { submissionRepository.deviceUIStateFlow } returns flow { emit(NetworkRequestWrapper.RequestIdle) }
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultAvailableFragmentTest.kt:50]: `flowOf()`?

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultAvailableFragmentTest.kt [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        every { submissionRepository.testResultReceivedDateFlow } returns flow { }
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultAvailableFragmentTest.kt:51]: MB here too?

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultFragmentTest.kt [hunk 4]
```diff
@@ -10,32 +10,68 @@ import androidx.test.ext.junit.runners.AndroidJUnit4
...
+import java.util.Date
 
 @RunWith(AndroidJUnit4::class)
 class SubmissionTestResultFragmentTest : BaseUITest() {
 
-    @MockK lateinit var pendingViewModel: SubmissionTestResultPendingViewModel
-    @MockK lateinit var uiState: TestResultUIState
+    lateinit var viewModel: SubmissionTestResultPendingViewModel
+    @MockK lateinit var submissionRepository: SubmissionRepository
+    @MockK lateinit var dispatcherProvider: DispatcherProvider
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultFragmentTest.kt:45]: TestDispatcher ?

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultNegativeFragmentTest.kt [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import testhelpers.SystemUIDemoModeRule
+import tools.fastlane.screengrab.locale.LocaleTestRule
+import java.util.Date
+
+@RunWith(AndroidJUnit4::class)
+class SubmissionTestResultNegativeFragmentTest : BaseUITest() {
+
+    lateinit var viewModel: SubmissionTestResultNegativeViewModel
+    @MockK lateinit var submissionRepository: SubmissionRepository
+    @MockK lateinit var dispatcherProvider: DispatcherProvider
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultNegativeFragmentTest.kt:36]: test dispatcher

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultNegativeFragmentTest.kt [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    @get:Rule
+    val systemUIDemoModeRule = SystemUIDemoModeRule()
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this, relaxed = true)
+
+        every { submissionRepository.deviceUIStateFlow } returns flow { emit(NetworkRequestWrapper.RequestIdle) }
+        every { submissionRepository.testResultReceivedDateFlow } returns flow { }
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultNegativeFragmentTest.kt:50]: flowOF

File: Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultNoConsentGivenFragmentTest.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import testhelpers.Screenshot
+import testhelpers.SystemUIDemoModeRule
+import tools.fastlane.screengrab.locale.LocaleTestRule
+import java.util.Date
+
+@RunWith(AndroidJUnit4::class)
+class SubmissionTestResultNoConsentGivenFragmentTest : BaseUITest() {
+
+    @MockK lateinit var submissionRepository: SubmissionRepository
+    @MockK lateinit var dispatcherProvider: DispatcherProvider
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionTestResultNoConsentGivenFragmentTest.kt:34]: same here

File: Corona-Warn-App/src/androidTest/java/testhelpers/BaseUITest.kt [hunk 8]
```diff
@@ -13,4 +15,11 @@ abstract class BaseUITest : BaseTest() {
     fun clearAllViewModels() {
         MockViewModelModule.CREATORS.clear()
     }
+
+    inline fun <reified F: Fragment>captureScreenshot(suffix:String = "") {
+        val name = F::class.simpleName
+        launchFragmentInContainer2<F>()
+        Thread.sleep(SCREENSHOT_DELAY_TIME)
+        Screengrab.screenshot(name.plus(suffix))
+    }
```
Comment (mtwalli) [inline Corona-Warn-App/src/androidTest/java/testhelpers/BaseUITest.kt:24]: 👍 , Please add it in `TestExtensions` file please consider also that in some situations , a different theme ID has to be passed to it. ex : captureScreenshot<Fragment>(R.style.AppTheme_Diary)

**Comment by mtwalli** [comment]
I'm getting this error message
```
[15:31:14]: ▸ de.rki.coronawarnapp.ui.submission.SubmissionTanFragmentTest:.INSTRUMENTATION_RESULT: shortMsg=Process crashed.
[15:31:14]: ▸ INSTRUMENTATION_CODE: 0
```
which prevents the submission flow from continuing for every locale except German

**Comment by jurajkusnier** [comment]
> I'm getting this error message
> 
> ```
> [15:31:14]: ▸ de.rki.coronawarnapp.ui.submission.SubmissionTanFragmentTest:.INSTRUMENTATION_RESULT: shortMsg=Process crashed.
> [15:31:14]: ▸ INSTRUMENTATION_CODE: 0
> ```
> 
> which prevents the submission flow from continuing for every locale except German
> 
> [Log.txt](https://github.com/corona-warn-app/cwa-app-android/files/5809247/Log.txt)

Can anybody else reproduce this issue? It's working on my machine.

**Comment by mtwalli** [comment]
> > I'm getting this error message
> > ```
> > [15:31:14]: ▸ de.rki.coronawarnapp.ui.submission.SubmissionTanFragmentTest:.INSTRUMENTATION_RESULT: shortMsg=Process crashed.
> > [15:31:14]: ▸ INSTRUMENTATION_CODE: 0
> > ```
> > 
> > 
> > which prevents the submission flow from continuing for every locale except German
> > [Log.txt](https://github.com/corona-warn-app/cwa-app-android/files/5809247/Log.txt)
> 
> Can anybody else reproduce this issue? It's working on my machine.

it is fine on Pixel 4a running API 30

**Review by mtwalli — DISMISSED** [review]
👍🏼

**Comment by ralfgehrer** [comment]
Looks good overall. When running it locally on an Emulator (Pixel2, Android 10), I noticed some differences between languages.
- There is a blank white screen vs. a home screen
- TAN entry is only available for some languages
- The "Your consent" screen before the QR Code is missing
(- The calendar in the symptoms screen is missing (not sure if we actually have to cover it - iOS does not))

Any idea if this could be a local problem or a general one?

Screenshot 2021-01-14 at 13 06 20

**Comment by jurajkusnier** [comment]
> Looks good overall. When running it locally on an Emulator (Pixel2, Android 10), I noticed some differences between languages.
> 
> * There is a blank white screen vs. a home screen
> * TAN entry is only available for some languages
> * The "Your consent" screen before the QR Code is missing
>   (- The calendar in the symptoms screen is missing (not sure if we actually have to cover it - iOS does not))
> 
> Any idea if this could be a local problem or a general one?
> 
> Screenshot 2021-01-14 at 13 06 20

- I didn't change the implementation of onboarding screenshots (white/empty screens), it may be problem with performance, we could try to change sleep time, but it looks like local problem. 
- Could you try to run it on real device? It may be also emulator issue as it's working ok my devices
- I'm going to check missing screens

**Comment by ralfgehrer** [comment]
Everything runs smoothly an actual device. Thanks!

With "The "Your consent" screen before the QR Code is missing" I intended to refer to the 
`SubmissionConsentFragment` instead of the `SubmissionYourConsentFragment`. Sorry for the confusion.

**Comment by jurajkusnier** [comment]
> Everything runs smoothly on an actual device. Many thanks!
> 
> With "The "Your consent" screen before the QR Code is missing" I intended to refer to the
> `SubmissionConsentFragment` instead of the `SubmissionYourConsentFragment`. Sorry for the confusion.

Sorry, `SubmissionConsentFragment` is also there now

**Review by ralfgehrer — APPROVED** [review]

**Review by mtwalli — APPROVED** [review]

**Pull request merged**